### PR TITLE
install_prereq: Fix package so asterisk can install on debian buster,bullseye, and bookworm.

### PR DESCRIPTION
**Pull request by jackdpeterson**
- [x] update list of debian packages to ones that are installable on debian:bullseye
- [x] switch from aptitude to apt-get (so we can get specific error messages when packages aren't found)
- [x] switch for aptitude check to dpkg since that's also built into the docker debian OS image

Other notes: this is a pretty messy script as-is and should probably have those constants split by version of OS. For example, debian bullseye doesn't have the mysql dependency available -- that's replaced w/ mariaDB equivalent client connector library. Furthermore, certain legacy libraries are replaced by newer versions.

### CODE REVIEW

**Review by gtjoseph — CHANGES_REQUESTED** [review]
A few things...
* Please open a corresponding "improvement" issue and add a "Fixes" or 'Resolves" line to the commit message per [Commit Messages](https://docs.asterisk.org/Development/Policies-and-Procedures/Commit-Messages/)
* Is this really docker specific?  Doesn't seem so.
* Update the commit message summary line so it's clear that it's only install_prereq that's changing:
`install_prereq: Update for debian:bullseye` might be better.
* Add the cherry-pick comment per [Code-Contribution](https://docs.asterisk.org/Development/Policies-and-Procedures/Code-Contribution/)
* You might want to list the packages that were actually changed in the commit message.
* How far back in ubuntu and debian versions are the changes still good?  If this breaks installing on ubuntu:18 or debian:buster for instance, this would be a problem.

Remember to follow the procedure for force pushing an amended commit described in [Code-Contribution](https://docs.asterisk.org/Development/Policies-and-Procedures/Code-Contribution/).
After you re-push up the commit, don't forget to change the PR description.  It doesn't update automatically.

**Comment by asteriskteam** [comment]
This PR has been marked stale because it has been in "Changes Requested" or "submitter-action-required" state for 28 days or more. Please make the requested changes within 14 days or the PR will be closed.

**Comment by jackdpeterson** [comment]
✅  Please open a corresponding "improvement" issue and add a "Fixes" or 'Resolves" line to the commit message per Is 

**this really docker specific? Doesn't seem so.**
Correct, updated messaging to reflect that this isn't docker-specific. That's just the environment I was testing in.

✅ Update the commit message summary line so it's clear that it's only install_prereq that's changing

✅  Add the cherry-pick comment per [Code-Contribution](https://docs.asterisk.org/Development/Policies-and-Procedures/Code-Contribution/)

✅ You might want to list the packages that were actually changed in the commit message.
-- Added package differences into commit.

How far back in ubuntu and debian versions are the changes still good? If this breaks installing on ubuntu:18 or debian:buster for instance, this would be a problem.
📆 -- Need to test this.

File: contrib/scripts/install_prereq [hunk 1]
```diff
@@ -190,9 +189,11 @@ in_test_mode() {
 
 check_installed_debs() {
 	for pack in "$@" ; do
-		tocheck="${tocheck} ^${pack}$ ~P^${pack}$"
+		EXISTS=`dpkg -l | cut -d ' ' -f 3 | grep ^"${pack}"\\$ | wc -l`
+		if [ ${EXISTS} -eq 0 ]; then
+		  echo "${pack}"
+    fi
```
Comment (seanbright) [inline contrib/scripts/install_prereq:195]:
```
if ! dpkg -l | awk 'NR > 5 {print $2}' | grep -q ^"${pack}"$; then
    echo "${pack}"
fi
```

Comment (seanbright) [reply contrib/scripts/install_prereq:195]: Actually, it may be as simple as:

`dpkg -l "${pack}" || echo "${pack}"`

**Review by seanbright — CHANGES_REQUESTED** [review]
You're going to have to switch between package lists based on the Debian/Ubuntu version. `lsb_release -a` gives you some of this info (at least on Ubuntu, I don't know about Debian).

**Comment by asteriskteam** [comment]
This PR has been marked stale because it has been in "Changes Requested" or "submitter-action-required" state for 28 days or more. Please make the requested changes within 14 days or the PR will be closed.

**Comment by jackdpeterson** [comment]
@seanbright -- I'd appreciate your feedback on this PR. I've modified my changes to separate out the different debian releases and select the packages there. This should make this more maintainable going forward and could be used as a pattern for other distributions that change/modify/remove packages over time.

In-scope: debian based OS's.
- [x] test debian:buster (10)
- [x]  test debain:bullseye (11)
- [x] test debian:bookworm (12)

[Dockerfiles I used to test](https://gist.github.com/jackdpeterson/c75eed4ac6e842e173fde80b56af3b04)

Commands to build:
## Buster
`docker build -t asterisk:buster -f Dockerfile-debian-buster .`

## Bullseye
`docker build -t asterisk:bullseye -f Dockerfile-debian-bullseye .`

`docker build -t asterisk:bookworm -f Dockerfile-debian-bookworm .`

Initially I tried compiling this on a Macbook Pro w/ an M1 chip; and compilation of asterisk failed there. What did work was launching a C5.2xlarge instance on AWS EC2 running Ubuntu 22.04 w/ git and docker.io installed.

File: contrib/scripts/install_prereq (hunk 1 above)
Comment (jackdpeterson) [reply contrib/scripts/install_prereq:247]: Hey @seanbright -- your suggestion here isn't quite safe:

```
cat test.sh

#!/usr/bin/env sh

PACK="SOMETHING_I_DONT_HAVE"

EXISTS=`dpkg -l "${pack}" >/dev/null 2>&1 || echo "${pack}"`
if [ ${EXISTS} -eq 0 ]; then
  echo "${pack}"
fi

./test.sh
./test.sh: 6: [: -eq: unexpected operator


```

The approach I have is guaranteed to provide a numeric output with the `| wc -l`

**Comment by jackdpeterson** [comment]
Could use a little bit of guidance on the failing tests here. Not making any changes to PJSIP.

**Comment by jcolp** [comment]
The install_prereq script is not used, the failing tests can be ignored for this PR.

**Comment by jackdpeterson** [comment]
Fixing one more issue; ensuring that dpkg -l section properly gets converted to a number always w/ `wc -l`
```
#24 1.203 ./contrib/scripts/install_prereq: 245: [: -eq: unexpected operator
#24 1.208 ./contrib/scripts/install_prereq: 245: [: -eq: unexpected operator
#24 1.213 ./contrib/scripts/install_prereq: 245: [: -eq: unexpected operator
#24 1.218 ./contrib/scripts/install_prereq: 245: [: Illegal number: libjansson-dev
```

File: contrib/scripts/install_prereq (hunk 1 above)
Comment (seanbright) [reply contrib/scripts/install_prereq]: I think you misunderstood my suggestion. I was suggesting this:

```shell
check_installed_debs() {
  for pack in "$@" ; do
    dpkg -l "${pack}" >/dev/null 2>&1 || echo "${pack}"
  done
}
```

File: contrib/scripts/install_prereq [hunk 2]
```diff
@@ -257,13 +310,25 @@ check_installed_zypper() {
...
+    extra_packs=`check_installed_debs $PACKAGES_DEBIAN_BUSTER`
+  fi
+
+  if [ $(cat /etc/debian_version | cut -d '.' -f 1) -eq 11 ]; then
+      extra_packs=`check_installed_debs $PACKAGES_DEBIAN_BULLSEYE`
+  fi
+
+  if [ $(cat /etc/debian_version | cut -d '.' -f 1) -eq 12 ]; then
+    extra_packs=`check_installed_debs $PACKAGES_DEBIAN_BOOKWORM`
+  fi
```
Comment (seanbright) [inline contrib/scripts/install_prereq:327]: Could this be simplified with a `case` statement? Something like:

```shell
case "$(cut -f1 -d. /etc/debian_version)" in
  10) extra_packs=$(check_installed_debs $PACKAGES_DEBIAN_BUSTER) ;;
  11) extra_packs=$(check_installed_debs $PACKAGES_DEBIAN_BULLSEYE) ;;
  12) extra_packs=$(check_installed_debs $PACKAGES_DEBIAN_BOOKWORM) ;;
  *)  extra_packs=$(check_installed_debs $PACKAGES_DEBIAN) ;;
esac
```

**Review by seanbright — CHANGES_REQUESTED** [review]

File: contrib/scripts/install_prereq (hunk 2 above)
Comment (jackdpeterson) [reply contrib/scripts/install_prereq:327]: That's pretty. I'll update the PR and test it to reflect that concept.

**Comment by asteriskteam** [comment]
This PR has been marked stale because it has been in "Changes Requested" or "submitter-action-required" state for 28 days or more. Please make the requested changes within 14 days or the PR will be closed.

**Comment by asteriskteam** [comment]
This PR has been closed because it has been in "Changes Requested" or "submitter-action-required" state for more than 42 days.

**Pull request closed without merging**